### PR TITLE
fix2640: Distributor refactoring to assert ingestion rate limits as early as possible

### LINKED ISSUE #2640: Distributor perf: assert ingestion rate limits as early as possible
Ingestion rate limits are asserted in the [distributor](https://github.com/grafana/tempo/blob/main/modules/distributor/distributor.go#L323-L331) after the received traces have been [parsed by the receiver](https://github.com/grafana/tempo/blob/main/modules/distributor/receiver/shim.go#L280-L295), [marshalled](https://github.com/grafana/tempo/blob/main/modules/distributor/distributor.go#L276-L281) and [unmarshalled](https://github.com/grafana/tempo/blob/main/modules/distributor/distributor.go#L283-L289) back. This results in a unnecessary CPU and mem usage if the batch is finally dropped.

Ingestion rate limits should be moved up, so there's as little processing as possible before asserting them. The receivers are OTel components, so changing them to accept our rate limits seems more complex than desired. We prefer to keep with the upstream version of this components. What seems the best choice is at the top of the [`PushTraces`](https://github.com/grafana/tempo/blob/main/modules/distributor/distributor.go#L272) function.

### PR DESCRIPTION
**Pull request by mghildiy**
**What this PR does**: Refactor distributor to ensure that rate limits are checked before other costly operations are done

**Which issue(s) this PR fixes**:
Fixes #2640

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 ## v2.2.0-rc0 / 2023-07-21
 
+* [CHANGE] Assert ingestion rate limits as early as possible [#2640](https://github.com/grafana/tempo/pull/2703) (@mghildiy)
```
Comment (joe-elliott) [inline CHANGELOG.md:5]: this change will not be in 2.2. We should move it above the # v2.2.0-rc0 line

File: modules/distributor/distributor.go [hunk 2]
```diff
@@ -268,11 +268,45 @@ func (d *Distributor) stopping(_ error) error {
...
+			overrides.ErrorPrefixRateLimited,
+			int(d.ingestionRateLimiter.Limit(now, userID)),
+			tracesSize)
+	}
+
+	return nil
+}
+
+func (d *Distributor) extractBasicInfo(ctx context.Context, traces ptrace.Traces) (userID string, spanCount, tracesSize int, err error) {
+	user, e := user.ExtractOrgID(ctx)
```
Comment (joe-elliott) [inline modules/distributor/distributor.go:286]: if this errors we should just return early instead of calling the next method

**Review by joe-elliott — APPROVED** [review]
Nice improvement :+1:

**Comment by joe-elliott** [comment]
looks like we're all fighting for space in the 2.3 changelog :). If you can resolve that conflict, I will merge.

**Comment by mghildiy** [comment]
> looks like we're all fighting for space in the 2.3 changelog :). If you can resolve that conflict, I will merge.

Done.

**Pull request merged**